File: native-lib/node/tests/tck/case-loader.ts
// Parses a TCK case directory into runnable scenarios, applying the same
// structural skip filters as the CLI's TCKCliTest.
//
// TCK case layout (one directory per case, defined by the runtime's
// FolderBasedTest):
//   transform.dwl      the transform (fixed main file name)
//   inN.<ext>          inputs; basename (in0, in1…) is the variable name,
//                      extension selects the reader format
//   out.<ext>          expected output; one scenario per out.* file
//   config.properties  optional per-case config
//
// The parsing here is pure — it operates on a provided list of file names, not
// the filesystem — so it is unit-testable without real TCK fixtures.
import { isSupportedExtension, mimeForExtension } from "./formats";

/** Fixed name of the transform script in a TCK case (per FolderBasedTest). */
export const MAIN_TRANSFORM = "transform.dwl";

const INPUT_PATTERN = /^in[0-9]+\.[a-zA-Z]+$/;
const OUTPUT_PATTERN = /^out\.[a-zA-Z]+$/;
const INPUT_CONFIG_PATTERN = /^in[0-9]+-config\.properties$/;
const OUTPUT_CONFIG_PATTERN = /^out[0-9]*-config\.properties$/;

/** Returns the extension (without dot, lowercased) of a file name, or "" if none. */
export function extensionOf(name: string): string {
  const i = name.lastIndexOf(".");
  return i < 0 ? "" : name.slice(i + 1).toLowerCase();
}

/** An input file within a case: the DataWeave variable name and its source file. */
export interface TckInput {
  /** Variable name the input binds to (the file's base name, e.g. `in0`). */
  name: string;
  /** The input file name (e.g. `in0.json`). */
  fileName: string;
  /** MIME type derived from the extension. */
  mimeType: string;
}

/** One runnable scenario: the transform plus its inputs and a single expected output. */
export interface TckScenario {
  /** Scenario id: `<case>-<outputFileName>`. */
  name: string;
  inputs: TckInput[];
  /** Expected-output file name (e.g. `out.json`). */
  outputFileName: string;
  /** Comparison format, from the output extension. */
  outputMime: string;
  outputExtension: string;
}

/** The result of inspecting one case directory. */
export type CaseParseResult =
  | { kind: "scenarios"; scenarios: TckScenario[] }
  | { kind: "skipped"; reason: string };

/**
 * Parses a case directory (given its file listing) into scenarios, or a skip
 * decision. Mirrors the CLI's structural filters: a case is skipped unless it
 * has exactly one transform.dwl, and is skipped outright if it carries
 * per-input/output config properties, groovy/java cases, a bare
 * config.properties, or a _wip marker. Scenarios whose input or output
 * extension maps to an unsupported format (e.g. yaml) are dropped; if none
 * remain the case is skipped.
 *
 * @param caseName - The directory (case) name, used as the scenario prefix.
 * @param fileNames - The file names directly inside the case directory.
 * @returns Either the runnable scenarios or a reason the case was skipped.
 */
export function parseCase(caseName: string, fileNames: string[]): CaseParseResult {
  if (caseName.endsWith("_wip") || caseName.endsWith("wip")) {
    return { kind: "skipped", reason: "work-in-progress (_wip)" };
  }

  if (fileNames.some((n) => INPUT_CONFIG_PATTERN.test(n) || OUTPUT_CONFIG_PATTERN.test(n))) {
    return { kind: "skipped", reason: "per-input/output config.properties not supported" };
  }
  if (fileNames.some((n) => n === "config.properties")) {
    return { kind: "skipped", reason: "config.properties not supported" };
  }
  if (fileNames.some((n) => n.endsWith(".groovy"))) {
    return { kind: "skipped", reason: "java/groovy case not supported" };
  }

  // Exactly one transform: a .dwl that is not an inN.dwl / out.dwl.
  const dwlFiles = fileNames.filter(
    (n) => extensionOf(n) === "dwl" && !INPUT_PATTERN.test(n) && !OUTPUT_PATTERN.test(n)
  );
  if (dwlFiles.length !== 1) {
    return { kind: "skipped", reason: `expected exactly one transform dwl, found ${dwlFiles.length}` };
  }
  if (!fileNames.includes(MAIN_TRANSFORM)) {
    return { kind: "skipped", reason: `transform is not named ${MAIN_TRANSFORM}` };
  }

  const inputFiles = fileNames.filter((n) => INPUT_PATTERN.test(n)).sort();
  const outputFiles = fileNames.filter((n) => OUTPUT_PATTERN.test(n)).sort();
  if (outputFiles.length === 0) {
    return { kind: "skipped", reason: "no expected output (out.*) file" };
  }

  // Drop scenarios that reference an unsupported input or output format.
  const unsupportedInput = inputFiles.find((n) => !isSupportedExtension(extensionOf(n)));
  if (unsupportedInput) {
    return { kind: "skipped", reason: `unsupported input format: ${unsupportedInput}` };
  }

  const inputs: TckInput[] = inputFiles.map((fileName) => ({
    name: fileName.slice(0, fileName.lastIndexOf(".")),
    fileName,
    mimeType: mimeOrThrow(fileName),
  }));

  const scenarios: TckScenario[] = outputFiles
    .filter((out) => isSupportedExtension(extensionOf(out)))
    .map((outputFileName) => {
      const outputExtension = extensionOf(outputFileName);
      return {
        name: `${caseName}-${outputFileName}`,
        inputs,
        outputFileName,
        outputExtension,
        outputMime: mimeOrThrow(outputFileName),
      };
    });

  if (scenarios.length === 0) {
    return { kind: "skipped", reason: "no scenarios with a supported output format" };
  }
  return { kind: "scenarios", scenarios };
}

function mimeOrThrow(fileName: string): string {
  // Guarded by isSupportedExtension at call sites; this keeps types non-null.
  const mime = mimeForExtension(extensionOf(fileName));
  if (!mime) throw new Error(`no MIME mapping for ${fileName}`);
  return mime;
}

File: native-lib/python/tests/tck/case_loader.py
"""Filesystem loader for the Gradle-staged DataWeave TCK corpus."""

from dataclasses import dataclass
from pathlib import Path
import re
from typing import Dict, List, Optional

from dataweave import InputValue


EXTENSION_TO_MIME = {
    "bin": "application/octet-stream",
    "csv": "application/csv",
    "dwl": "application/dw",
    "json": "application/json",
    "multipart": "multipart/form-data",
    "properties": "text/x-java-properties",
    "txt": "text/plain",
    "urlencoded": "application/x-www-form-urlencoded",
    "xml": "application/xml",
}

INPUT_PATTERN = re.compile(r"^in[0-9]+\.[a-zA-Z]+$")
OUTPUT_PATTERN = re.compile(r"^out\.[a-zA-Z]+$")
INPUT_CONFIG_PATTERN = re.compile(r"^in[0-9]+-config\.properties$")
OUTPUT_CONFIG_PATTERN = re.compile(r"^out[0-9]*-config\.properties$")


@dataclass(frozen=True)
class TckScenario:
    identifier: str
    transform: str
    inputs: Dict[str, InputValue]
    expected: bytes
    output_extension: str
    charset: Optional[str]


@dataclass(frozen=True)
class DiscoveredCase:
    identifier: str
    scenarios: List[TckScenario]


@dataclass(frozen=True)
class Discovery:
    cases: List[DiscoveredCase]
    structural_skips: int
    structural_case_identifiers: List[str]
    structural_module_case_identifiers: List[str]


def extension_of(name: str) -> str:
    return Path(name).suffix.removeprefix(".").lower()


def discover_cases(suites_dir: Path) -> Discovery:
    cases: List[DiscoveredCase] = []
    structural_skips = 0
    structural_case_identifiers: List[str] = []
    structural_module_case_identifiers: List[str] = []
    if not suites_dir.exists():
        return Discovery(
            cases,
            structural_skips,
            structural_case_identifiers,
            structural_module_case_identifiers,
        )

    for suite_dir in sorted(path for path in suites_dir.iterdir() if path.is_dir()):
        for case_dir in sorted(path for path in suite_dir.iterdir() if path.is_dir()):
            if not case_dir.exists():
                structural_skips += 1
                structural_case_identifiers.append(f"{suite_dir.name}/{case_dir.name}")
                continue
            scenarios = _load_case(suite_dir.name, case_dir)
            if scenarios is None:
                structural_skips += 1
                identifier = f"{suite_dir.name}/{case_dir.name}"
                structural_case_identifiers.append(identifier)
                if _has_adjacent_dwl_module(case_dir):
                    structural_module_case_identifiers.append(identifier)
            else:
                cases.append(DiscoveredCase(f"{suite_dir.name}/{case_dir.name}", scenarios))
    return Discovery(
        cases,
        structural_skips,
        structural_case_identifiers,
        structural_module_case_identifiers,
    )


def _has_adjacent_dwl_module(case_dir: Path) -> bool:
    return any(
        path.is_file()
        and extension_of(path.name) == "dwl"
        and not INPUT_PATTERN.fullmatch(path.name)
        and not OUTPUT_PATTERN.fullmatch(path.name)
        and path.name != "transform.dwl"
        for path in case_dir.iterdir()
    )


def _load_case(suite_name: str, case_dir: Path) -> Optional[List[TckScenario]]:
    files = {path.name: path for path in case_dir.iterdir() if path.is_file()}
    case_name = case_dir.name
    if case_name.endswith("_wip") or case_name.endswith("wip"):
        return None
    if "config.properties" in files or any(
        INPUT_CONFIG_PATTERN.fullmatch(name) or OUTPUT_CONFIG_PATTERN.fullmatch(name)
        for name in files
    ):
        return None
    if any(extension_of(name) == "groovy" for name in files):
        return None

    transforms = [
        path for name, path in files.items()
        if extension_of(name) == "dwl"
        and not INPUT_PATTERN.fullmatch(name)
        and not OUTPUT_PATTERN.fullmatch(name)
    ]
    if len(transforms) != 1 or "transform.dwl" not in files:
        return None

    input_paths = sorted(
        (path for name, path in files.items() if INPUT_PATTERN.fullmatch(name)),
        key=lambda path: path.name,
    )
    if any(extension_of(path.name) not in EXTENSION_TO_MIME for path in input_paths):
        return None
    output_paths = sorted(
        (
            path
            for name, path in files.items()
            if OUTPUT_PATTERN.fullmatch(name) and extension_of(name) in EXTENSION_TO_MIME
        ),
        key=lambda path: path.name,
    )
    if not output_paths:
        return None

    inputs = {
        path.stem: InputValue(path.read_bytes(), mime_type=EXTENSION_TO_MIME[extension_of(path.name)])
        for path in input_paths
    }
    transform = files["transform.dwl"].read_text(encoding="utf-8")
    encoding = files.get("encoding")
    charset = encoding.read_text(encoding="utf-8").strip() if encoding else None
    case_id = f"{suite_name}/{case_name}"
    return [
        TckScenario(
            identifier=f"{case_id}:{path.name}",
            transform=transform,
            inputs=inputs,
            expected=path.read_bytes(),
            output_extension=extension_of(path.name),
            charset=charset,
        )
        for path in output_paths
    ]
